test/functional/horizon_controller_test.go

```diff
@@ -0,0 +1,279 @@
+package functional_test
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	horizon "github.com/openstack-k8s-operators/horizon-operator/api/v1alpha1"
+	keystonev1 "github.com/openstack-k8s-operators/keystone-operator/api/v1beta1"
+	condition "github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+)
+
+const (
+	// TODO(gibi): Do we want to test in a realistic namespace like "openstack"?
+	TestNamespace = "default"
+
+	SecretName = "osp-secret"
+
+	timeout  = time.Second * 2
+	interval = time.Millisecond * 200
+)
+
+func DefaultHorizonTemplate() *horizon.Horizon {
+	return &horizon.Horizon{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "horizon.openstack.org/v1alpha1",
+			Kind:       "Horizon",
+		},
+		ObjectMeta: metav1.ObjectMeta{},
+		Spec: horizon.HorizonSpec{
+			ContainerImage: "test-horizon-container-image",
+			Secret:         SecretName,
+		},
+	}
+}
+
+type TestHorizon struct {
+	LookupKey types.NamespacedName
+	// The input data for creating a Horizon
+	Template *horizon.Horizon
+	// The current state of the Horizon resource updated by Refresh()
+	Instance *horizon.Horizon
+}
+
+// NewTestHorizon initializes the the input for a Horizon instance
+// but does not create it yet. So the client can finetuned the Template data
+// before calling Create()
+func NewTestHorizon(namespace string) TestHorizon {
+	name := fmt.Sprintf("horizon-%s", uuid.New().String())
+	template := DefaultHorizonTemplate()
+	template.ObjectMeta.Name = name
+	template.ObjectMeta.Namespace = namespace
+	return TestHorizon{
+		LookupKey: types.NamespacedName{Name: name, Namespace: namespace},
+		Template:  template,
+		Instance:  &horizon.Horizon{},
+	}
+}
+
+// Creates the Horizon resource in k8s based on the Template. The Template
+// is not updated during create. This call waits until the resource is created.
+// The last known state of the resource is available via Instance.
+func (t TestHorizon) Create() {
+	Expect(k8sClient.Create(ctx, t.Template.DeepCopy())).Should(Succeed())
```

Comment:
One thing I learned in the hard way was that using golang structs for creating resources has a non trivial side effect. When you say horizon.Horizon{} you get a full go struct with all the fields have some a value. If you don't provide a field in the template then go will initialize it with the null value regardless if you have kubebilder annotations with default values. Then Create here will serialize the go stuct to json. This way you cannot really simulate that some field is not provided in the create request to let the API server do the defaulting during deserialization of the json payload. So in nova we switched to unstructured Create to be able to test with poper default values.  https://github.com/openstack-k8s-operators/nova-operator/blob/e6742035a36d7c865620511d25d5784feb721999/test/functional/base_test.go#L52-L86